distutils/compilers/C/msvc.py

```diff
@@ -371,17 +371,24 @@ def out_extensions(self) -> dict[str, str]:
 
     def compile(  # noqa: C901
         self,
-        sources,
+        sources: Sequence[str | os.PathLike[str]],
         output_dir=None,
         macros=None,
         include_dirs=None,
         debug=False,
         extra_preargs=None,
         extra_postargs=None,
         depends=None,
-    ):
+    ) -> list[str]:
         if not self.initialized:
             self.initialize()
+
+        # Re-sort the list of sources so that .mc files come first
+        # See pypa/setuptools#4986
+        sources = sorted(
+            sources, key=lambda source: (not str(source).endswith(".mc"), source)
```

Comment:
A tiny bit more complex, but I could look for the extensions in `_mc_extensions` instead of "hardocding" `.mc`.

I could also extract this to a function to better comment what they is doing:
The "key" is a tuple to do grouped sorting. The first element is `0` for mc files, which is sorted before `1` for everything else.
The second element is the original path to keep the same sorting as in `build_ext`.